MoreLinq/FallbackIfEmpty.cs

```diff
@@ -162,45 +161,45 @@ public static IEnumerable<T> FallbackIfEmpty<T>(this IEnumerable<T> source, IEnu
             int? count, T fallback1, T fallback2, T fallback3, T fallback4,
             IEnumerable<T> fallback)
         {
-            var collection = source as ICollection<T>;
-            if (collection != null && collection.Count == 0)
-            {
-                //
-                // Replace the empty collection with an empty sequence and
-                // carry on. LINQ's Enumerable.Empty is implemented
-                // intelligently to return the same enumerator instance and so
-                // does not incur an allocation. However, the same cannot be
-                // said for a collection like an empty array or list. This
-                // permits the rest of the logic while keeping the call to
-                // source.GetEnumerator() cheap.
-                //
-
-                source = Enumerable.Empty<T>();
+            if (source is ICollection<T> collection) {
+                if (collection.Count == 0) {
+                    return fallbackChooser();
+                }
+                else {
+                    return collection;
+                }
             }
 
-            using (var e = source.GetEnumerator())
+            return nonCollectionIterator();
+
+            IEnumerable<T> nonCollectionIterator()
             {
-                if (e.MoveNext())
-                {
-                    do { yield return e.Current; }
-                    while (e.MoveNext());
-                }
-                else
-                {
-                    e.Dispose(); // eager disposal
-                    if (count > 0 && count <= 4)
-                    {
-                        yield return fallback1;
-                        if (count > 1) yield return fallback2;
-                        if (count > 2) yield return fallback3;
-                        if (count > 3) yield return fallback4;
-                    }
-                    else
-                    {
-                        foreach (var item in fallback)
-                            yield return item;
+                using (var e = source.GetEnumerator()) {
+                    if (e.MoveNext()) {
+                        do { yield return e.Current; }
+                        while (e.MoveNext());
+                        yield break;
                     }
                 }
+
+                foreach (var item in fallbackChooser())
+                    yield return item;
+            }
+            IEnumerable<T> fallbackChooser()
+            {
+                if (count > 0 && count <= 4) {
```

Comment:
I see. Do we want to add the assertions? I guess what bothers me is having to worry about the range and seeing the 4 hard-coded. This is one way to get rid of that:

```c#
IEnumerable<T> FallbackChooser()
{
    return count == null ? fallback : InstancesFallback(count.Value);

    IEnumerable<T> InstancesFallback(int c)
    {
        if (c < 0) throw new ArgumentOutOfRangeException(nameof(count));
        if (c > 0) { c--; yield return fallback1; }
        if (c > 0) { c--; yield return fallback2; }
        if (c > 0) { c--; yield return fallback3; }
        if (c > 0) { c--; yield return fallback4; }
        if (c > 0) throw new ArgumentOutOfRangeException(nameof(count));
    }
}
```

The downside is that the last throw won't happen until after the last yield so yet another way is:

```c#
IEnumerable<T> FallbackChooser()
{
    switch (count)
    {
        case null: return fallback;
        case int n when n == 1: return Fallback1();
        case int n when n == 2: return Fallback2();
        case int n when n == 3: return Fallback3();
        case int n when n == 4: return Fallback4();
        default: throw new ArgumentOutOfRangeException(nameof(count));
    }

    IEnumerable<T> Fallback1() { yield return fallback1; }
    IEnumerable<T> Fallback2() => Fallback1().Concat(fallback2);
    IEnumerable<T> Fallback3() => Fallback2().Concat(fallback3);
    IEnumerable<T> Fallback4() => Fallback3().Concat(fallback4);
}
```